### PR TITLE
Updating format of doubles in VCFs.

### PR DESCRIPTION
**Pull request by nh13**
@tfenne a rough implementation of https://github.com/samtools/htsjdk/issues/793.  I can probably make the function itself faster.

Related to https://github.com/samtools/htsjdk/issues/793

### CODE REVIEW

File: src/main/java/htsjdk/variant/vcf/VCFEncoder.java [hunk 1]
```diff
@@ -201,33 +198,66 @@ else if ( val instanceof List ) {
 		return result;
 	}
 
+	private static double round(final double value, final int digits) {
```
Comment (yfarjoun) [inline src/main/java/htsjdk/variant/vcf/VCFEncoder.java]: should this be in some MathUtils class?

File: src/main/java/htsjdk/variant/vcf/VCFEncoder.java [hunk 2]
```diff
@@ -201,33 +198,66 @@ else if ( val instanceof List ) {
...
-				}
+		final int numSignificant = 5;
+		final double lowerLimit = Math.pow(10, -numSignificant);
+		if (d == 0.0) {
+			return "0.0";
+		}
+		else if (d == Double.POSITIVE_INFINITY || d == Double.NEGATIVE_INFINITY || Double.isNaN(d)) {
+			return String.format("%f", d);
+		}
+		else if (d >= 0.1) {
```
Comment (yfarjoun) [inline src/main/java/htsjdk/variant/vcf/VCFEncoder.java]: what about negative numbers?

File: src/test/java/htsjdk/variant/vcf/VCFEncoderTest.java [hunk 3]
```diff
@@ -21,38 +21,53 @@
...
-
-	@Test(dataProvider = "VCFWriterDoubleFormatTestData")
-	public void testVCFWriterDoubleFormatTestData(final double d, final String expected) {
-		Assert.assertEquals(VCFEncoder.formatVCFDouble(d), expected, "Failed to pretty print double in VCFWriter");
-	}
+    @DataProvider(name = "VCFWriterDoubleFormatTestData")
+    public Object[][] makeVCFWriterDoubleFormatTestData() {
+        final List<Object[]> tests = new ArrayList<Object[]>();
+        // New spec test cases
+        tests.add(new Object[]{0.0, "0.0"});
```
Comment (yfarjoun) [inline src/test/java/htsjdk/variant/vcf/VCFEncoderTest.java]: please add tests to negative numbers as well.

Comment (yfarjoun) [inline src/test/java/htsjdk/variant/vcf/VCFEncoderTest.java]: please add tests to negative numbers as well.

File: src/main/java/htsjdk/variant/vcf/VCFEncoder.java [hunk 4]
```diff
@@ -201,33 +198,66 @@ else if ( val instanceof List ) {
...
-					format = "%.3e";
-				else {
-					// return a zero format
-					return "0.00";
-				}
+		final int numSignificant = 5;
+		final double lowerLimit = Math.pow(10, -numSignificant);
+		if (d == 0.0) {
+			return "0.0";
+		}
```
Comment (yfarjoun) [inline src/main/java/htsjdk/variant/vcf/VCFEncoder.java]: `} else {` 
here and below

File: src/test/java/htsjdk/variant/vcf/VCFEncoderTest.java [hunk 5]
```diff
@@ -21,38 +21,53 @@
...
+        tests.add(new Object[]{0.01, "0.01"});
+        tests.add(new Object[]{0.001, "0.001"});
+        tests.add(new Object[]{0.0001, "0.0001"});
+        tests.add(new Object[]{0.00001, "0.00001"});
+        tests.add(new Object[]{0.000001, "1.00000e-06"});
+        tests.add(new Object[]{0.50000001, "0.5"});
+        tests.add(new Object[]{0.012345, "1.23450e-02"});
+        tests.add(new Object[]{0.0033333333, "3.33333e-03"});
+        // Previous test cases
+        tests.add(new Object[]{1.0, "1.0"});
```
Comment (yfarjoun) [inline src/test/java/htsjdk/variant/vcf/VCFEncoderTest.java]: please add tests to negative numbers as well.

**Review by yfarjoun — CHANGES_REQUESTED** [review]

**Comment by yfarjoun** [comment]
also, some tests are failing....please fix.

**Comment by yfarjoun** [comment]
hey @nh13 are you still interested in this change?

**Comment by nh13** [comment]
@yfarjoun assigned to myself, thanks!

File: src/main/java/htsjdk/variant/vcf/VCFEncoder.java [hunk 6]
```diff
@@ -201,33 +198,66 @@ else if ( val instanceof List ) {
 		return result;
 	}
 
+	private static double round(final double value, final int digits) {
+		final double x = Math.pow(10, digits);
+
+		return ((int)(value * x)) / x;
```
Comment (yfarjoun) [inline src/main/java/htsjdk/variant/vcf/VCFEncoder.java]: this should probbly use Math.round rather than (int) so that we do not always round down.

**Comment by fleharty** [comment]
@nh13 Just taking this over unless you object.

File: src/main/java/htsjdk/variant/vcf/VCFEncoder.java (hunk 1 above)
Comment (fleharty) [reply src/main/java/htsjdk/variant/vcf/VCFEncoder.java]: We had discussed using a MathUtils class for this, but none exists for rounding to a particular number of significant digits.  It's not clear to me that the current code correctly rounds to the number of significant digits as we would like it to.

I think we want a function that does something like:  (for digits = 4)
0.000003599999 -> 0.0000036
4.12345 -> 4.123
1234.345 -> 1234

**Comment by yfarjoun** [comment]
any movement on this Mark?

**Comment by fleharty** [comment]
@yfarjoun Trying to make time for this.  I plan to fix this up sometime later this week.

**Comment by nh13** [comment]
@fleharty no urgency on my end, but just checking in on this.

**Comment by yfarjoun** [comment]
@fleharty will you have time to finish this?

**Comment by fleharty** [comment]
@yfarjoun I've got a ton of other things going on at the moment, I should punt on this and let someone else finish.

**Comment by yfarjoun** [comment]
@fleharty I see that you removed your assignment...have you managed to make any progress on this PR before deciding not to pursue it?
